### PR TITLE
Initial s3 import

### LINKED ISSUE #4: Initial S3 download
### Description
S3logsbeat can read new elements added to S3 via SQS. However, old ones are not included to ElasticSearch.

Below is an option to implement this and can changes can be discussed.

### Configuration
Add a new input type called `s3` on yaml config file with these fields:
* `buckets`: array of string with format `s3://bucket_name/bucket_path`
* `since`: get only those objects with creation timestamp after or equal to this field. Ignore if this field is not set. Must be in ISO8601 format.
* `to`: get only those objects with creation timestamp before or equal to this field. Ignore if this field is not set. Must be in ISO8601 format.

Also must have these fields from `sqs` type:
* `log_format`
* `key_regex_fields`

Would be perfect if `ignore_keys` (from #3 ) is added to this new input type.

Add a new command called `s3import` to import these s3 logs. In this mode execution, `sqs` type present on yaml file are ignored and type `s3` is only taken into account when this command is executed and ignored in other cases.

### Execution
Execution could be something similar to:
* Command `s3import` is executed with a config file with types `s3` present on it.
* All S3 objects present on buckets and paths present on field `buckets` that matches timestamps between `since` and `to` are imported to ElasticSearch.
* Once all S3 objects are imported, the execution finishes.

### PR DESCRIPTION
**Pull request by mpucholblasco**
# Notes
Added initial S3 import.

This PR closes #4 .

More information about process on README.

### CODE REVIEW

File: aws/s3object.go [hunk 1]
```diff
@@ -1,6 +1,15 @@
...
-import "fmt"
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+var (
+	s3uriRE = regexp.MustCompile(`^s3://(?P<bucket>[^/]+)/(?P<key>.*)$`)
```
Comment (luismiramirez) [inline aws/s3object.go]: What does `RE` mean?

Comment (mpucholblasco) [reply aws/s3object.go]: `RE` -> `Regular Expression`

File: beater/s3importsbeat.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	wgEvents := &eventCounter{
+		count: monitoring.NewInt(nil, "s3logsbeat.events.active"),
+		added: monitoring.NewUint(nil, "s3logsbeat.events.added"),
+		done:  monitoring.NewUint(nil, "s3logsbeat.events.done"),
+		err:   monitoring.NewUint(nil, "s3logsbeat.events.parserError"),
+	}
+	finishedLogger := newFinishedLogger(wgEvents)
+
+	// Setup registrar to persist state
+	registrar := registrar.New(finishedLogger)
```
[the remark this answers is not in the document]
Comment (mpucholblasco) [reply beater/s3importsbeat.go]: no, this registrar is related to `libbeats` (the library used to connect to ElasticSearch). The registrar is present on folder `registrar` and it's used to register events with `libbeats`. In our case, it's used to know when an event is acknowledged by ElasticSearch.

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+require (
+	github.com/aws/aws-sdk-go v1.18.3
+	github.com/elastic/beats v6.6.2+incompatible
+	github.com/elastic/go-ucfg v0.7.0 // indirect
+	github.com/gofrs/uuid v3.2.0+incompatible // indirect
+	github.com/joeshaw/multierror v0.0.0-20140124173710-69b34d4ec901 // indirect
+	github.com/mitchellh/hashstructure v1.0.0
+	github.com/pkg/errors v0.8.1 // indirect
+	github.com/sequra/s3logsbeat v0.0.0-20190407090053-41cebb8188ae // indirect
```
[the remark this answers is not in the document]
Comment (mpucholblasco) [reply go.mod]: Yes, me neither. This is autogenerated by `go`.

File: input/s3/input.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	config    config
+	done      chan struct{}
+	out       chan *pipeline.S3List
+	logParser logparser.LogParser
+}
+
+// NewInput instantiates a new Log
+func NewInput(
+	cfg *common.Config,
+	context input.Context,
```
[the remark this answers is not in the document]
Comment (mpucholblasco) [reply input/s3/input.go]: What do you mean? Putting `context` first and the `cfg`?

File: input/s3/config.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package sqs
```
[the remark this answers is not in the document]
Comment (mpucholblasco) [reply input/s3/config.go]: copy & paste error, well seen!

File: input/s3/input.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
+package sqs
```
[the remark this answers is not in the document]
Comment (mpucholblasco) [reply input/s3/input.go]: copy & paste error, well seen!

File: beater/s3importsbeat.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+		pipelineChannels.GetS3ListChannel(),
+		[]string{"s3"},
+	)
+	if err != nil {
+		logp.Err("Could not init crawler: %v", err)
+		return err
+	}
+
+	// Start the registrar
+	registrar.Start()
```
Comment (bismarkhenao) [inline beater/s3importsbeat.go]: This *registrar* points to the variable created on the line 65 right? It's a bit confusing maybe? XD

File: input/input.go [hunk 8]
```diff
@@ -66,9 +69,10 @@ func New(
...
 	context := Context{
-		Done:     input.done,
-		BeatDone: input.beatDone,
-		Out:      input.out,
+		Done:      input.done,
+		BeatDone:  input.beatDone,
+		OutSQS:    input.outSQS,
+		OutS3List: input.outS3List,
 	}
 	var ipt Input
```
Comment (bismarkhenao) [inline input/input.go]: Should be all the declarations at the top of the block maybe? I know it does not matter here, I was thinking in the structure :)

File: beater/s3importsbeat.go (hunk 7 above)
Comment (bismarkhenao) [inline beater/s3importsbeat.go]: This registrar points to the variable created on the line 65 right? It's a bit confusing maybe? XD

File: input/s3/config.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	"time"
+
+	"github.com/sequra/s3logsbeat/input"
+)
+
+var (
+	defaultConfig = config{}
+)
+
+type config struct {
```
Comment (bismarkhenao) [inline input/s3/config.go]: Should be **Config** instead **config**? I think all the types you have created until now are PascalCase

File: input/s3/input.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bismarkhenao) [reply input/s3/input.go]: I think so

File: input/s3/config.go (hunk 9 above)
Comment (mpucholblasco) [reply input/s3/config.go]: No, in go lowercase means `private` to the package and CameCase means `public` (to be used externally). In this case, this config is only used locally.

File: input/input.go (hunk 8 above)
Comment (mpucholblasco) [reply input/input.go]: That's derived from C. In this case, it's better to declare when you gonna use (or event instantiate when you use it).

File: beater/s3importsbeat.go (hunk 7 above)
Comment (mpucholblasco) [reply beater/s3importsbeat.go]: It depends on dependencies used on constructor. In this case, we have to initialize all objects and then start to use (after all initialization started).

File: input/s3/config.go (hunk 9 above)
Comment (bismarkhenao) [reply input/s3/config.go]: Ok! :) 👌

**Pull request merged**